### PR TITLE
Improve `CreateDirectory` on Windows

### LINKED ISSUE #61954: Improve CreateDirectory on Windows
The CreateDirectory implementation was improved on Unix by https://github.com/dotnet/runtime/pull/58799 and https://github.com/dotnet/runtime/pull/61777.

The first PR eliminates exists syscalls by deducting that information from the directory creation syscall. The second PR reduces allocations when creating parent directories by allocating the parent paths on the stack instead of creating strings.

Similar improvements can be made to the Windows implementation.

cc @adamsitnik @danmoseley

### PR DESCRIPTION
**Pull request by deeprobin**
Fixes #61954

As described in https://github.com/dotnet/runtime/issues/61954#issuecomment-1069635649 this only provides minor improvements.
Here it must be clearly decided which commits we take over and which are reverted again (Decision between micro improvement and code readability).
Personally, I also find it unnecessary that we use a `List`. However, I have not yet found a better way to get around this.
Even if the benchmarks don't show a big improvement, the interop calls in case of an error have been reduced and so in case of a `throw` the performance is improved (even if this is not a hot path)

In any case, an improvement of the memory allocation can be seen, since a lot has been spanified (see commit https://github.com/dotnet/runtime/commit/6faa0e15b86d26e1ac912536b091d68ac3385247).

### Before
|          Method |     Mean |    Error |   StdDev |   Median |      Min |      Max | Allocated |
|---------------- |---------:|---------:|---------:|---------:|---------:|---------:|----------:|
| CreateDirectory | 195.6 us | 18.80 us | 20.90 us | 191.3 us | 174.4 us | 247.5 us |     497 B |

### After
|          Method |     Mean |   Error |   StdDev |   Median |      Min |      Max | Allocated |
|---------------- |---------:|--------:|---------:|---------:|---------:|---------:|----------:|
| CreateDirectory | 187.7 us | 9.26 us | 10.29 us | 186.1 us | 171.3 us | 204.9 us |     455 B |

/cc @danmoseley 
/cc @adamsitnik 
/cc @tmds

### CODE REVIEW

File: src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs [hunk 1]
```diff
@@ -83,10 +93,9 @@ public static unsafe void CreateDirectory(string fullPath, byte[]? securityDescr
                     lpSecurityDescriptor = (IntPtr)pSecurityDescriptor
                 };
 
-                while (stackDir.Count > 0)
+                for (int i = count - 1; 0 <= i; i--)
```
Comment (gfoidl) [inline src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:96]:
```suggestion
                for (int i = count - 1; i >= 0; i--)
```
Don't use yoda conditions here.

File: src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs [hunk 2]
```diff
@@ -40,6 +51,15 @@ public static bool FileExists(string fullPath)
                 ((data.dwFileAttributes & Interop.Kernel32.FileAttributes.FILE_ATTRIBUTE_DIRECTORY) == 0);
         }
 
+        public static (int fileAttributes, int lastError) GetFileAttributes(string fullPath)
```
Comment (danmoseley) [inline src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs:54]: this and DirectoryExistsSlim return last error in inconsistent ways.

can you just have a single method, that returns `(bool fileExists, bool directoryExists)` ? (or `(bool fileExists, bool directoryExists, int lastError)` if necessary). that would also make the callers simpler.

Comment (deeprobin) [reply src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs]: I had to remove this in commit 8cfca8641f11d052c8c81a1cb3a415104c59dd13 as it caused the test to fail.
But this part is only executed if the directory already exists.

**Comment by deeprobin** [comment]
*I have updated the benchmarks.*

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.FindFirstFileEx.cs [hunk 3]
```diff
@@ -20,6 +20,11 @@ internal static SafeFindHandle FindFirstFile(string fileName, ref WIN32_FIND_DAT
         {
             fileName = PathInternal.EnsureExtendedPrefixIfNeeded(fileName);
 
+            // use FindExInfoBasic since we don't care about short name and it has better perf
+            return FindFirstFilePrefixed(fileName, ref data);
+        }
+        internal static SafeFindHandle FindFirstFilePrefixed(string fileName, ref WIN32_FIND_DATA data)
```
Comment (danmoseley) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.FindFirstFileEx.cs:26]: nit - new line before }

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetFileAttributesEx.cs [hunk 4]
```diff
@@ -21,6 +21,10 @@ internal static partial class Kernel32
         internal static bool GetFileAttributesEx(string? name, GET_FILEEX_INFO_LEVELS fileInfoLevel, ref WIN32_FILE_ATTRIBUTE_DATA lpFileInformation)
         {
             name = PathInternal.EnsureExtendedPrefixIfNeeded(name);
+            return GetFileAttributesExPrefixed(name, fileInfoLevel, ref lpFileInformation);
+        }
+        internal static bool GetFileAttributesExPrefixed(string? name, GET_FILEEX_INFO_LEVELS fileInfoLevel, ref WIN32_FILE_ATTRIBUTE_DATA lpFileInformation)
```
Comment (danmoseley) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetFileAttributesEx.cs:26]: newline

File: src/libraries/Common/src/System/IO/PathInternal.cs [hunk 5]
```diff
@@ -241,6 +241,12 @@ internal static bool RemoveRelativeSegments(ReadOnlySpan<char> path, int rootLen
         internal static bool EndsInDirectorySeparator(ReadOnlySpan<char> path) =>
             path.Length > 0 && IsDirectorySeparator(path[path.Length - 1]);
 
+        /// <summary>
+        /// Same as <see cref="EndsInDirectorySeparator(string?)"/> but without length check.
+        /// </summary>
+        internal static bool EndsInDirectorySeparatorUnchecked(ReadOnlySpan<char> path) =>
```
Comment (danmoseley) [inline src/libraries/Common/src/System/IO/PathInternal.cs:247]: I suggest to just inline both of EndsInDirectorySeparatorUnchecked and EndsInDirectorySeparator since they are so simple and only exist in 3 or 4 places. then you can remove these methods.

**Comment by deeprobin** [comment]
> github won't let me add this comment, but
> 
> > EndsWithPeriodOrSpace and EndsWithPeriodOrSpaceSlim are each only essentially used in 1 place in the tree now and are both simple. I would inline them both and remove the methods.

I think there are problems on the part of GitHub. I just could not adjust the description of another PR.

grafik

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetFileAttributesEx.cs (hunk 4 above)
Comment (danmoseley) [reply src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetFileAttributesEx.cs:26]: also should this method be inlined? It's just a pass through now.

File: src/libraries/Common/src/System/IO/PathInternal.Windows.cs [hunk 6]
```diff
@@ -72,11 +72,11 @@ internal static bool IsValidDriveChar(char value)
...
         }
 
-        internal static bool EndsWithPeriodOrSpace(string? path)
-        {
-            if (string.IsNullOrEmpty(path))
-                return false;
+        internal static bool EndsWithPeriodOrSpace(string? path) =>
+            !string.IsNullOrEmpty(path) && EndsWithPeriodOrSpaceSlim(path);
 
+        internal static bool EndsWithPeriodOrSpaceSlim(string path)
```
Comment (danmoseley) [inline src/libraries/Common/src/System/IO/PathInternal.Windows.cs:78]: Looks like this method can be turned into a single line and produce the same code:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHufuBLAHYYYUAdgA2TckgbAIESQEEAFMXIAGBgAdsGABYBKLrx6dqJi0wDs23XoDaO/QDoAMjAEBzfQzgNyALoMALzBDADkEQyEhLb6jnZuHt56vv5BoRHO4QDcxhYAvvkMxbyCwqISUjJyCgwAQqoacYalpm0WYHrYUAxgIS0JLu5ePn6BeeaWvMQ2/ZmRkTF9IWHh2ZPTRdQFQA===
(checked both 32 and 64 bit)

Given that, I suggest inlining `path[path.Length - 1] == ' ' || path[path.Length - 1] == '.'` into EndsWithPeriodOrSpace and EnsureExtendedPrefixIfNeeded. This eliminates a method that isn't really useful, but also is a trap for anyone in future that calls it without checking the length. It's perfectly clear when inlined.

Comment (danmoseley) [reply src/libraries/Common/src/System/IO/PathInternal.Windows.cs:78]: BTW, we have slightly changed the behavior here. Previously EndsWithPeriodOrSpace would return false for "      " and now it will return true. Is that OK? I'm actually not sure.

If it is, then EndsWithPeriodOrSpace could be even less code: `return path.Length > 0 && (path[path.Length - 1] == ' ' || path[path.Length - 1] == '.');` generates less code. But, I don't suggest changing that here.

File: src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs [hunk 7]
```diff
@@ -49,53 +60,64 @@ public static bool FileExists(string fullPath)
...
+        /// <summary>
+        /// Same as <see cref="FillAttributeInfo"/> without separator-trimming and media-insertion blocking
+        /// </summary>
+        /// <param name="path">The file path from which the file attribute information will be filled.</param>
+        /// <param name="data">A struct that will contain the attribute information.</param>
+        /// <param name="returnErrorOnNotFound">Return the error code for not found errors?</param>
+        internal static int FillAttributeInfoSlim(string? path, ref Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA data, bool returnErrorOnNotFound)
+        {
+            int errorCode = Interop.Errors.ERROR_SUCCESS;
+            string? prefixedString = PathInternal.EnsureExtendedPrefixIfNeeded(path);
```
Comment (danmoseley) [inline src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs:81]: The FindFirstFilePrefixed path here is the rare case, right? (<1%) Where the file is locked.

Given that, I don't think you're really saving anything by ensuring you only call EnsureExtendedPrefixIfNeeded once. I would reverse this and leave GetFileAttributesExPrivate private. That avoids creating a way for future code to accidentally call GetFileAttributesEx without prefixing. And it leaves this code a bit simpler.

File: src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs [hunk 8]
```diff
@@ -49,53 +60,64 @@ public static bool FileExists(string fullPath)
...
+            if (!Interop.Kernel32.GetFileAttributesExPrivate(prefixedString, Interop.Kernel32.GET_FILEEX_INFO_LEVELS.GetFileExInfoStandard, ref data))
+            {
+                errorCode = Marshal.GetLastWin32Error();
+
+                Interop.Kernel32.WIN32_FIND_DATA findData = default;
+                if (!IsPathUnreachableError(errorCode))
                 {
-                    errorCode = Marshal.GetLastWin32Error();
+                    // Assert so we can track down other cases (if any) to add to our test suite
+                    Debug.Assert(errorCode is Interop.Errors.ERROR_ACCESS_DENIED or Interop.Errors.ERROR_SHARING_VIOLATION or Interop.Errors.ERROR_SEM_TIMEOUT,
```
Comment (danmoseley) [inline src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs:91]: Please don't change style like this in the same PR -- we should generally leave existing code in the same style as other nearby code. If we need to change it, we would do that en-masse in its own PR.

File: src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs [hunk 9]
```diff
@@ -34,39 +34,49 @@ public static unsafe void CreateDirectory(string fullPath, byte[]? securityDescr
...
 
             bool somepathexists = false;
-            int length = fullPath.Length;
+
+            ReadOnlySpan<char> fullPathSpan = fullPath;
+            int length = fullPathSpan.Length;
 
             // We need to trim the trailing slash or the code will try to create 2 directories of the same name.
-            if (length >= 2 && PathInternal.EndsInDirectorySeparator(fullPath.AsSpan()))
+            if (length >= 2 && PathInternal.IsDirectorySeparator(fullPathSpan[^1]))
```
Comment (danmoseley) [inline src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:42]: does it save anything to avoid calling AsSpan() twice? I would have thought that is very fast.

File: src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs [hunk 10]
```diff
@@ -34,39 +34,49 @@ public static unsafe void CreateDirectory(string fullPath, byte[]? securityDescr
...
             {
                 length--;
             }
 
-            int lengthRoot = PathInternal.GetRootLength(fullPath.AsSpan());
+            int lengthRoot = PathInternal.GetRootLength(fullPathSpan);
 
-            if (length > lengthRoot)
+            using (DisableMediaInsertionPrompt.Create())
```
Comment (danmoseley) [inline src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:49]: why add this -- was it missing?

File: src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs [hunk 11]
```diff
@@ -121,11 +130,12 @@ public static unsafe void CreateDirectory(string fullPath, byte[]? securityDescr
             // Handle CreateDirectory("X:\\") when X: doesn't exist. Similarly for n/w paths.
             if ((count == 0) && !somepathexists)
             {
-                string? root = Path.GetPathRoot(fullPath);
+                ReadOnlySpan<char> root = Path.GetPathRoot(fullPathSpan);
```
Comment (danmoseley) [inline src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:133]: why change this? you end up creating the string either way.

it might be nice if DirectoryExists overload existed that took a span, to save creating it, but not suggesting that's worth doing here. This is a rare code path, right?

File: src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs (hunk 10 above)
Comment (deeprobin) [reply src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:49]: This is one of the main reasons why we use the Slim Method.
In the non-Slim method, we do this at the method level.

DisableMediaInsertionPrompt makes a kernel32 call at initialization and dispose, which we execute only once instead of calling `DirectoryExistsSlim`.

https://github.com/dotnet/runtime/blob/c926642522a3cc9723a7056f4f730ec2a39bf210/src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs#L54-L70

File: src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs (hunk 9 above)
Comment (deeprobin) [reply src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:42]: I actually assumed that this would change something. 
But this intrinsic is probably optimized well enough.

Of course, I find the introduction of a local more readable. 

A good rule of thumb is, if you know the output of a function is deterministic and you use that output multiple times, make a local output for it!

@danmoseley 
In that case, however, we have no performance improvement from it (CodeGen is the same).

Should I remove this?

Comment (deeprobin) [reply src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:42]: @danmoseley friendly ping :smile:

Comment (danmoseley) [reply src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:42]: oops. Yes, I'd keep it as was. AsSpan() is idiomatic - we use it in the hottest places.

File: src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs (hunk 11 above)
Comment (deeprobin) [reply src/libraries/Common/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:133]: It's not worth it here in my opinion because this calls a WinAPI method (interop with spans doesn't work, does it?).

File: src/libraries/Common/src/System/IO/PathInternal.Windows.cs (hunk 6 above)
Comment (deeprobin) [reply src/libraries/Common/src/System/IO/PathInternal.Windows.cs:78]: We have only two direct usages, which could break:

### PathInternal.EnsureExtendedPrefixIfNeeded
```cs
        [return: NotNullIfNotNull("path")]
        internal static string? EnsureExtendedPrefixIfNeeded(string? path)
        {
            if (path != null && (path.Length >= MaxShortPath || EndsWithPeriodOrSpaceSlim(path)))
            {
                return EnsureExtendedPrefix(path);
            }
            else
            {
                return path;
            }
        }
    ```

and

### FileSystemInfo.NormalizedPath
```cs
        internal string NormalizedPath
            => PathInternal.EndsWithPeriodOrSpace(FullPath) ? PathInternal.EnsureExtendedPrefix(FullPath) : FullPath;
```

@tmds @Jozkee Are you comfortable with this change (Tests are passing, but behavior might slightly change)?

Comment (jeffhandley) [reply src/libraries/Common/src/System/IO/PathInternal.Windows.cs:78]: @deeprobin I'd like to err on the side of caution here and avoid the subtle behavior change.

**Comment by jeffhandley** [comment]
@deeprobin I got this branch caught back up with main and reran the checks; there are some tests failing related to the long paths and the affected code. On my local Windows 11 machine, when I set `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem\LongPathsEnabled` to `0`, the failing tests reproduce locally.

```
~\git\dotnet\runtime\src\libraries\System.IO.FileSystem\tests [issue-61954-rev ≡] # dotnet test --filter "System.IO.Tests.DirectoryInfo_Create.DirectoryLongerThanMaxDirectoryAsPath_Succeeds"
  Determining projects to restore...
  All projects are up-to-date for restore.
  Microsoft.Interop.SourceGeneration -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\Microsoft.Interop.SourceGeneration\Debug\netstandard2.0\Microsoft.Interop.SourceGeneration.dll
  LibraryImportGenerator -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\LibraryImportGenerator\Debug\netstandard2.0\Microsoft.Interop.LibraryImportGenerator.dll
  TestUtilities -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\TestUtilities\Debug\net6.0\TestUtilities.dll
  System.Diagnostics.EventLog.Messages -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.Diagnostics.EventLog.Messages\Debug\netstandard2.0\System.Diagnostics.EventLog.Messages.dll
  System.Runtime -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.Runtime\ref\Debug\net7.0\System.Runtime.dll
  System.Security.Claims -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.Security.Claims\ref\Debug\net7.0\System.Security.Claims.dll
  System.Collections.NonGeneric -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.Collections.NonGeneric\ref\Debug\net7.0\System.Collections.NonGeneric.dll
  System.Security.Principal.Windows -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.Security.Principal.Windows\ref\Debug\net7.0\System.Security.Principal.Windows.dll
  System.Diagnostics.EventLog -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.Diagnostics.EventLog\ref\Debug\net7.0\System.Diagnostics.EventLog.dll
  System.Security.AccessControl -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.Security.AccessControl\ref\Debug\net7.0\System.Security.AccessControl.dll
  System.IO.FileSystem.AccessControl -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.IO.FileSystem.AccessControl\ref\Debug\net7.0\System.IO.FileSystem.AccessControl.dll
  System.IO.FileSystem.AccessControl -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.IO.FileSystem.AccessControl\Debug\net7.0-windows\System.IO.FileSystem.AccessControl.dll
  System.Diagnostics.EventLog -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.Diagnostics.EventLog\Debug\net7.0-windows\System.Diagnostics.EventLog.dll
  System.ServiceProcess.ServiceController -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.ServiceProcess.ServiceController\ref\Debug\net7.0\System.ServiceProcess.ServiceController.dll
  System.ServiceProcess.ServiceController -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.ServiceProcess.ServiceController\Debug\net7.0-windows\System.ServiceProcess.ServiceController.dll
  StreamConformanceTests -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\StreamConformanceTests\Debug\net7.0\StreamConformanceTests.dll
  System.IO.FileSystem.Tests -> C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.IO.FileSystem.Tests\Debug\net7.0-windows\System.IO.FileSystem.Tests.dll
Test run for C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.IO.FileSystem.Tests\Debug\net7.0-windows\System.IO.FileSystem.Tests.dll (.NETCoreApp,Version=v7.0)
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:01:05.87]     System.IO.Tests.DirectoryInfo_Create.DirectoryLongerThanMaxDirectoryAsPath_Succeeds [FAIL]
  Failed System.IO.Tests.DirectoryInfo_Create.DirectoryLongerThanMaxDirectoryAsPath_Succeeds [478 ms]
  Error Message:
   Assert.All() Failure: 3 out of 3 items in the collection did not pass.
[2]: Item: C:\Users\jeffhand\AppData\Local\Temp\#DirectoryInfo_Create_lwiaguzz.mxc\DirectoryLongerThanMaxDirectoryAsPath_Succeeds_293_b25p5hmt\686c58d928f441b285b1de79365f0e77gggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg
     System.IO.PathTooLongException: The path 'C:\Users\jeffhand\AppData\Local\Temp\#DirectoryInfo_Create_lwiaguzz.mxc\DirectoryLongerThanMaxDirectoryAsPath_Succeeds_293_b25p5hmt\686c58d928f441b285b1de79365f0e77gggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg' is too long, or a component of the specified path is too long.
        at System.IO.FileSystem.CreateDirectory(String fullPath, Byte[] securityDescriptor) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\Common\src\System\IO\FileSystem.DirectoryCreation.Windows.cs:line 145
        at System.IO.DirectoryInfo.Create() in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.Private.CoreLib\src\System\IO\DirectoryInfo.cs:line 89
        at System.IO.Tests.DirectoryInfo_Create.Create(String path) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.IO.FileSystem\tests\DirectoryInfo\Create.cs:line 15
        at System.IO.Tests.Directory_CreateDirectory.<DirectoryLongerThanMaxDirectoryAsPath_Succeeds>b__29_0(String path) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.IO.FileSystem\tests\Directory\CreateDirectory.cs:line 296
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/xunit.assert/Asserts/CollectionAsserts.cs:line 34
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/xunit.assert/Asserts/CollectionAsserts.cs:line 61
[1]: Item: C:\Users\jeffhand\AppData\Local\Temp\#DirectoryInfo_Create_lwiaguzz.mxc\DirectoryLongerThanMaxDirectoryAsPath_Succeeds_293_b25p5hmt\cc4253b47386458684bd29ae0f88d49fggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg
     System.IO.PathTooLongException: The path 'C:\Users\jeffhand\AppData\Local\Temp\#DirectoryInfo_Create_lwiaguzz.mxc\DirectoryLongerThanMaxDirectoryAsPath_Succeeds_293_b25p5hmt\cc4253b47386458684bd29ae0f88d49fggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg' is too long, or a component of the specified path is too long.
        at System.IO.FileSystem.CreateDirectory(String fullPath, Byte[] securityDescriptor) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\Common\src\System\IO\FileSystem.DirectoryCreation.Windows.cs:line 145
        at System.IO.DirectoryInfo.Create() in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.Private.CoreLib\src\System\IO\DirectoryInfo.cs:line 89
        at System.IO.Tests.DirectoryInfo_Create.Create(String path) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.IO.FileSystem\tests\DirectoryInfo\Create.cs:line 15
        at System.IO.Tests.Directory_CreateDirectory.<DirectoryLongerThanMaxDirectoryAsPath_Succeeds>b__29_0(String path) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.IO.FileSystem\tests\Directory\CreateDirectory.cs:line 296
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/xunit.assert/Asserts/CollectionAsserts.cs:line 34
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/xunit.assert/Asserts/CollectionAsserts.cs:line 61
[0]: Item: C:\Users\jeffhand\AppData\Local\Temp\#DirectoryInfo_Create_lwiaguzz.mxc\DirectoryLongerThanMaxDirectoryAsPath_Succeeds_293_b25p5hmt\214cfc26777e4378876d8ed508d2fee7gggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg
     System.IO.PathTooLongException: The path 'C:\Users\jeffhand\AppData\Local\Temp\#DirectoryInfo_Create_lwiaguzz.mxc\DirectoryLongerThanMaxDirectoryAsPath_Succeeds_293_b25p5hmt\214cfc26777e4378876d8ed508d2fee7gggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggggg' is too long, or a component of the specified path is too long.
        at System.IO.FileSystem.CreateDirectory(String fullPath, Byte[] securityDescriptor) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\Common\src\System\IO\FileSystem.DirectoryCreation.Windows.cs:line 145
        at System.IO.DirectoryInfo.Create() in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.Private.CoreLib\src\System\IO\DirectoryInfo.cs:line 89
        at System.IO.Tests.DirectoryInfo_Create.Create(String path) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.IO.FileSystem\tests\DirectoryInfo\Create.cs:line 15
        at System.IO.Tests.Directory_CreateDirectory.<DirectoryLongerThanMaxDirectoryAsPath_Succeeds>b__29_0(String path) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.IO.FileSystem\tests\Directory\CreateDirectory.cs:line 296
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/xunit.assert/Asserts/CollectionAsserts.cs:line 34
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/xunit.assert/Asserts/CollectionAsserts.cs:line 61
  Stack Trace:
     at System.IO.Tests.Directory_CreateDirectory.DirectoryLongerThanMaxDirectoryAsPath_Succeeds() in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.IO.FileSystem\tests\Directory\CreateDirectory.cs:line 294
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr) in C:\Users\jeffhand\git\dotnet\runtime\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodInvoker.cs:line 64
Results File: C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.IO.FileSystem.Tests\Debug\net7.0-windows\TestResults\jeffhand_JEFFHAND-DESK_2022-06-30_23_56_51.trx
Html test results file : C:\Users\jeffhand\git\dotnet\runtime\artifacts\bin\System.IO.FileSystem.Tests\Debug\net7.0-windows\TestResults\TestResult_jeffhand_JEFFHAND-DESK_20220630_235651.html

Failed!  - Failed:     1, Passed:     0, Skipped:     0, Total:     1, Duration: 15 ms - System.IO.FileSystem.Tests.dll (net7.0)
```

**Comment by deeprobin** [comment]
I take a look

File: src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs [hunk 12]
```diff
@@ -49,53 +60,64 @@ public static bool FileExists(string fullPath)
...
+                    // Another case where enumeration works is with special system files such as
+                    // pagefile.sys that give back ERROR_SHARING_VIOLATION on GetAttributes.
+                    //
+                    // Ideally we'd only try again for known cases due to the potential performance
+                    // hit. The last attempt to do so baked for nearly a year before we found the
+                    // pagefile.sys case. As such we're probably stuck filtering out specific
+                    // cases that we know we don't want to retry on.
 
-                    if (!IsPathUnreachableError(errorCode))
+                    using SafeFindHandle handle = Interop.Kernel32.FindFirstFile(prefixedString!, ref findData);
```
Comment (stephentoub) [inline src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs:107]: What's the purpose of this change?

File: src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs [hunk 13]
```diff
@@ -29,6 +29,17 @@ private static bool DirectoryExists(string? path, out int lastError)
...
+        private static bool DirectoryExistsSlim(string? path, out int lastError)
+        {
+            Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA data = default;
+            lastError = FillAttributeInfoSlim(path, ref data, returnErrorOnNotFound: true);
+
+            return
+                (lastError == 0) &&
+                (data.dwFileAttributes != -1) &&
+                ((data.dwFileAttributes & Interop.Kernel32.FileAttributes.FILE_ATTRIBUTE_DIRECTORY) != 0);
+        }
```
Comment (stephentoub) [inline src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs:41]: Adding these "Slim" functions seems to be an attempt to improve throughput, but the perf results shared don't really show a throughput improvement.  Seems like these changes should be reverted?

File: src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs (hunk 7 above)
Comment (stephentoub) [reply src/libraries/Common/src/System/IO/FileSystem.Attributes.Windows.cs:81]: This was marked resolved but I don't see the feedback addressed.  Was it commented on somewhere else?

**Review by stephentoub — COMMENTED** [review]
Thanks for working on this, but I'm unclear at this time what improvements the PR is concretely providing.  Where are the cited allocation improvements coming from?

**Comment by danmoseley** [comment]
I'm going to close this as it's unclear what the benefits are. Feel free to reopen if you address feedback and can show evidence of improvement making the change worthwhile. Thanks.
